src/sage/rings/semirings/tropical_polynomial.py

```diff
@@ -0,0 +1,1016 @@
+
+        Note that the factor `x - x_0` in classical algebra gets transformed
+        to the factor `x + x_0`, since the root of the tropical polynomial
+        `x + x_0` is `x_0` and not `-x_0`. However, not every tropical
+        polynomial can be factored.
+
+        OUTPUT: a :class:'Factorization'
+
+        EXAMPLES::
+
+            sage: T = TropicalSemiring(QQ, use_min=True)
+            sage: R.<x> = PolynomialRing(T)
+            sage: p1 = R([6,3,1,0]); p1
+            0*x^3 + 1*x^2 + 3*x + 6
+            sage: factor(p1)
+            (0) * (0*x + 1) * (0*x + 2) * (0*x + 3)
+
+        Such factorization is not always possible::
+
+            sage: p2 = R([4,4,2]); p2
+            2*x^2 + 4*x + 4
+            sage: p2.factor()
+            (2) * (0*x^2 + 2*x + 2)
+
+        TESTS:
+
+        The factorization for a constant::
+
+            sage: p3 = R(3)
+            sage: p3.factor()
+            (3) * 0
+        """
+        from sage.structure.factorization import Factorization
+        unit = self.dict()[self.degree()]
+        if (self != self.split_form()) or (not self.roots()):
+            factor = [(self * self.parent(-unit.lift()), 1)]
+            return Factorization(factor, unit=unit)
+
+        R = self.parent()
+        roots_order = {}
+        for root in self.roots():
+            if root in roots_order:
+                roots_order[root] += 1
+            else:
+                roots_order[root] = 1
+        factors = []
+        for root in roots_order:
+            factors.append((R([root, 0]), roots_order[root]))
+
+        return Factorization(factors, unit=unit)
+
+    def piecewise_function(self):
+        r"""
+        Return the tropical polynomial function of ``self``.
+
+        The function is a piecewise linear function with the domains are
+        divided by the roots. First we convert each term of polynomial to
+        its corresponding linear function. Next, we must determine which
+        term achieves the minimum (maximum) at each interval.
+
+        OUTPUT: A piecewise function
+
+        EXAMPLES::
+
+            sage: T = TropicalSemiring(QQ, use_min=False)
+            sage: R.<x> = PolynomialRing(T)
+            sage: p1 = R([4,2,1,3]); p1
+            3*x^3 + 1*x^2 + 2*x + 4
+            sage: p1.piecewise_function()
+            piecewise(x|-->4 on (-oo, 1/3], x|-->3*x + 3 on (1/3, +oo); x)
+
+        A constant tropical polynomial will result in a constant function::
+
+            sage: p2 = R(3)
+            sage: p2.piecewise_function()
+            3
+
+        A monomial will resulted in a linear function::
+
+            sage: p3 = R(1)*x^3
+            sage: p3.piecewise_function()
+            3*x + 1
+        """
+        from sage.symbolic.ring import SR
+        from sage.functions.piecewise import piecewise
+        from sage.sets.real_set import RealSet
+
+        x = SR.var('x')
+        if not self.roots():
```

Comment:
```suggestion
        x = SR.var('x')
        data = self.dict()
        if not self.roots():
```
and use `data` below instead of repeatedly calling `self.dict()`.